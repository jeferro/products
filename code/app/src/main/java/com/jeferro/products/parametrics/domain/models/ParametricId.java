package com.jeferro.products.parametrics.domain.models;

import com.jeferro.shared.ddd.domain.models.aggregates.StringIdentifier;

public class ParametricId extends StringIdentifier {

    public ParametricId(String value) {
        super(value);
    }

    public static ParametricId createOf(String value) {
        return new ParametricId(value);
    }
}
