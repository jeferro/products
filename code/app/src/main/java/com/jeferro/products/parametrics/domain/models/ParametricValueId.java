package com.jeferro.products.parametrics.domain.models;

import com.jeferro.shared.ddd.domain.models.aggregates.StringIdentifier;

public class ParametricValueId extends StringIdentifier {

    public ParametricValueId(String value) {
        super(value);
    }

    public static ParametricValueId createOf(String value) {
        return new ParametricValueId(value);
    }
}
