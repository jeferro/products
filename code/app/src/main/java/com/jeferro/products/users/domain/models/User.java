package com.jeferro.products.users.domain.models;

import java.util.Set;

import com.jeferro.products.shared.domain.exceptions.ValueValidationException;
import com.jeferro.products.shared.domain.models.aggregates.AggregateRoot;
import com.jeferro.products.shared.domain.models.users.Username;
import org.apache.commons.lang3.StringUtils;
import org.springframework.util.CollectionUtils;

public class User extends AggregateRoot<Username> {

    private final String encodedPassword;

    private final Set<String> roles;

    public User(Username username, String encodedPassword, Set<String> roles) {
        super(username);

        validateEncodedPassword(encodedPassword);
        validateRoles(roles);

        this.encodedPassword = encodedPassword;
        this.roles = roles;
    }

    public Username getUsername() {
        return id;
    }

    public String getEncodedPassword() {
        return encodedPassword;
    }

    public Set<String> getRoles() {
        return roles;
    }

	private static void validateEncodedPassword(String encodedPassword) {
		if (StringUtils.isBlank(encodedPassword)) {
			throw ValueValidationException.createOfMessage("Encoded password is blank");
		}
	}

	private static void validateRoles(Set<String> roles) {
		if (CollectionUtils.isEmpty(roles)) {
			throw ValueValidationException.createOfMessage("Roles is empty");
		}
	}
}
