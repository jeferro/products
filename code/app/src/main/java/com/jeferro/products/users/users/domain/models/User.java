package com.jeferro.products.users.users.domain.models;

import java.util.Set;

import com.jeferro.shared.ddd.domain.models.aggregates.AggregateRoot;
import com.jeferro.shared.ddd.domain.models.auth.Username;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;

public class User extends AggregateRoot<Username> {

  private String encodedPassword;

  private Set<String> roles;

  public User(Username id, String encodedPassword, Set<String> roles) {
	super(id);

	setEncodedPassword(encodedPassword);
	setRoles(roles);
  }

  public Username getUsername() {
	return id;
  }

  public String getEncodedPassword() {
	return encodedPassword;
  }

  public Set<String> getRoles() {
	return roles;
  }

  private void setEncodedPassword(String encodedPassword) {
	ValueValidationUtils.isNotBlank(encodedPassword, "Encoded password");
	this.encodedPassword = encodedPassword;
  }

  private void setRoles(Set<String> roles) {
	ValueValidationUtils.isNotEmpty(roles, "Roles");
	this.roles = roles;
  }
}
