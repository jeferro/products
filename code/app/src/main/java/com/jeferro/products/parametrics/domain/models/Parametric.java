package com.jeferro.products.parametrics.domain.models;

import com.jeferro.products.parametrics.domain.exceptions.ParametricValueNotFoundException;
import com.jeferro.products.parametrics.domain.models.values.ParametricValueId;
import com.jeferro.products.parametrics.domain.models.values.ParametricValues;
import com.jeferro.shared.ddd.domain.models.projection.Projection;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import com.jeferro.shared.locale.domain.models.LocalizedField;
import lombok.Getter;

@Getter
public class Parametric extends Projection<ParametricId> {

  private LocalizedField name;

  private ParametricValues values;

  public Parametric(
	  ParametricId id,
	  LocalizedField name,
	  ParametricValues values) {
	super(id);

	setName(name);
	setValues(values);
  }

  public static Parametric createOf(ParametricId id, LocalizedField name, ParametricValues values) {
	return new Parametric(id, name, values);
  }

  public boolean validate(ParametricValueId parametricValueId) {
	if (notContainsParametricValue(parametricValueId)) {
	  throw ParametricValueNotFoundException.createOf(this, parametricValueId);
	}

	return true;
  }

  private boolean notContainsParametricValue(ParametricValueId parametricValueId) {
	return values.stream()
		.noneMatch(parametricValue -> parametricValue.hasSameId(parametricValueId));
  }

  public void setName(LocalizedField name) {
	ValueValidationUtils.isNotNull(name, "name", this);
	this.name = name;
  }

  private void setValues(ParametricValues values) {
	ValueValidationUtils.isNotEmpty(values, "values", this);
	this.values = values;
  }
}
