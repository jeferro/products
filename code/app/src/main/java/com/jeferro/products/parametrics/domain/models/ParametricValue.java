package com.jeferro.products.parametrics.domain.models;

import java.util.List;

import com.jeferro.shared.ddd.domain.models.projection.Projection;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import com.jeferro.shared.locale.domain.models.LocalizedField;
import lombok.Getter;

@Getter
public class ParametricValue extends Projection<ParametricValueId> {

  private LocalizedField name;

  private String value;

  private List<ParametricValue> values;

  public ParametricValue(
	  ParametricValueId id,
	  LocalizedField name,
	  String value,
	  List<ParametricValue> values) {
	super(id);

	setName(name);
	setValue(value);
	setValues(values);
  }

  public static ParametricValue createSimple(ParametricValueId id, LocalizedField name) {
	return new ParametricValue(id, name, null, null);
  }

  private void setName(LocalizedField name) {
	ValueValidationUtils.isNotNull(name, "name", this);
	this.name = name;
  }

  private void setValue(String value) {
	this.value = value;
  }

  private void setValues(List<ParametricValue> values) {
	this.values = values;
  }
}
