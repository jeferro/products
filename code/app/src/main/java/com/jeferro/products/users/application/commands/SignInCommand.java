package com.jeferro.products.users.application.commands;

import com.jeferro.shared.application.commands.Command;
import com.jeferro.products.shared.domain.exceptions.internals.ValueValidationException;
import com.jeferro.shared.domain.models.auth.Auth;
import com.jeferro.shared.domain.models.auth.Username;
import com.jeferro.products.users.domain.models.User;
import org.apache.commons.lang3.StringUtils;

public class SignInCommand extends Command<User> {

    private Username username;

    private String plainPassword;

    public SignInCommand(Auth auth, Username username, String plainPassword) {
        super(auth);

		setUsername(username);
		setPlainPassword(plainPassword);
    }

	public Username getUsername() {
        return username;
    }

    public String getPlainPassword() {
        return plainPassword;
    }

	private void setUsername(Username username) {
		if (username == null) {
			throw ValueValidationException.createOfMessage("Username is null");
		}

	  this.username = username;
	}

	private void setPlainPassword(String plainPassword) {
		if (StringUtils.isBlank(plainPassword)) {
			throw ValueValidationException.createOfMessage("Plain password is blank");
		}

	  this.plainPassword = plainPassword;
	}
}
