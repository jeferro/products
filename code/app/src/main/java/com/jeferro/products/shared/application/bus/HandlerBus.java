package com.jeferro.products.shared.application.bus;

import com.jeferro.products.shared.application.Handler;
import com.jeferro.products.shared.application.commands.Command;
import com.jeferro.products.shared.domain.exceptions.internals.InternalErrorException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

public abstract class HandlerBus {

    private static final Logger logger = LoggerFactory.getLogger(HandlerBus.class);

    private final Map<Class<Command<?>>, Handler<?, ?>> handlers = new HashMap<>();

    public <R> R execute(Command<R> command) {
        var handler = getHandler(command);

        if (handler == null) {
            logger.error("Handler not found by command {}", command.getClass().getSimpleName());

            throw InternalErrorException.createOfHandlerNotFound();
        }

        return handler.execute(command);
    }

    protected void registryHandler(Handler<?, ?> handler) {
        Type type = handler.getClass().getGenericSuperclass();

        if (!(type instanceof ParameterizedType parameterizedType)) {
            throw new IllegalArgumentException("Handler superclass is not a parameterized type");
        }

        Class<Command<?>> commandClass = (Class<Command<?>>) parameterizedType.getActualTypeArguments()[0];

        handlers.put(commandClass, handler);
    }

    private <R> Handler<Command<R>, R> getHandler(Command<R> command) {
        Class<?> commandClass = command.getClass();

        if (!handlers.containsKey(commandClass)) {
            return null;
        }

        return (Handler<Command<R>, R>) handlers.get(commandClass);
    }
}
