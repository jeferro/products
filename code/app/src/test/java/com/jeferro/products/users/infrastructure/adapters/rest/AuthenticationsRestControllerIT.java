package com.jeferro.products.users.infrastructure.adapters.rest;

import com.jeferro.products.shared.application.bus.HandlerBus;
import com.jeferro.products.shared.infrastructure.adapters.rest.RestControllerIT;
import com.jeferro.products.shared.infrastructure.adapters.utils.ApprovalUtils;
import com.jeferro.products.users.application.commands.SignInCommand;
import com.jeferro.products.users.domain.models.User;
import com.jeferro.products.users.domain.models.UserMother;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import static org.mockito.Mockito.when;

@WebMvcTest(AuthenticationsRestController.class)
class AuthenticationsRestControllerIT extends RestControllerIT {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private HandlerBus handlerBus;


    @Test
    void givenAnUsers_whenSignIn_thenExecutesSignInCommand() throws Exception {
        var user = UserMother.user();
        ArgumentCaptor<SignInCommand> commandCaptor = givenAnAuthenticatedUserOnUserSignIn(user);

        var requestContent = """
                {
                  "username": "%s",
                  "password": "plain-password"
                }"""
                .formatted(user.getUsername());

        var requestBuilder = MockMvcRequestBuilders.post("/v1/authentications")
                .contentType(MediaType.APPLICATION_JSON)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(commandCaptor.getValue(),
                response.getStatus(),
                response.getContentAsString());
    }


    private ArgumentCaptor<SignInCommand> givenAnAuthenticatedUserOnUserSignIn(User user) {
        ArgumentCaptor<SignInCommand> commandCaptor = ArgumentCaptor.forClass(SignInCommand.class);

        when(handlerBus.execute(commandCaptor.capture()))
                .thenReturn(user);

        return commandCaptor;
    }
}