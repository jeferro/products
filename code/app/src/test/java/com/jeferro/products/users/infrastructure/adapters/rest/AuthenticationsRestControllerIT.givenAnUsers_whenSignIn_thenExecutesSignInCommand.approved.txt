params[0] = SignInCommand[plainPassword=plain-password,username=user,auth=anonymous]
params[1] = 200
params[2] = {"username":"user","roles":["user"]}
